### ISSUE TITLE
DJ tips

### ISSUE THREAD

**Issue by mdrxy** [WBOR-91-1-FM/wbor-studio-dashboard#3]
Popping up now and then (like a video game loading screen)

Could perhaps have a subcategory of tips applicable to new DJs only?

**Comment by mdrxy** [comment]
Perhaps instead of pop-ups, they could take on a loading-screen-esque form and just be in constant rotation in some segment of the screen.

**Comment by CaspianA1** [comment]
I think having constant DJ tips might not be ideal since eventually, new DJs would exhaust the full list of tips over an hour in the studio

Using the surprise system in place would work rly well with this
The tips would then just be a smaller window, and maybe every ~15 minutes (on average), a new tip pops up for ~30 seconds

**Comment by CaspianA1** [comment]
Also tips for contributing to MGMT (joining, or working on dashboard)

**Comment by CaspianA1** [comment]
Once this is done, remove the `tips.txt` file

**Comment by mdrxy** [comment]
Why remove the file? I think they'd be sourced from a file rather than hard-coded.

**Comment by CaspianA1** [comment]
In a way, it does make sense to do this. It's just that this feels like a bit of a slippery slope.

If we put the tips in a file, then what else should we put in a file?

In an ideal world, dashboard themes would be loadable from some format like JSON.
But since the fields for what defines a dashboard "theme" are not standardized, they're kept in the files `standard.rs`, `barebones.rs`, etc.

So I'm trying to keep as much data as possible to load each theme in the `.rs` files, for now. In a way, those files are both code and data, and fragmenting the data typically kept in those files to be an external asset feels like it breaks this philosophy.

Technical note - my eventual plan is to allow surprises to display text textures as well - so I'd add the tips into the list [here](https://github.com/WBOR-91-1-FM/wbor-studio-dashboard/blob/1c21aa63fd3170131c29d36ad27827d9cb730a60/src/dashboard_defs/themes/barebones.rs#L235).

**Comment by mdrxy** [comment]
I'm not seeing how tips for DJs are related to themes here?

**Comment by CaspianA1** [comment]
A lot of code is repeated between each `.rs` file that loads themes - in fact, most of the code between each file is repeated. In each theme's `.rs` file, there's a list of surprises to load (all of which are currently the same for each theme). The current tactic is to allow for this repetition; so adding DJ tips would mean 1. adding support for text textures in surprises, and 2. inserting DJ tips into each theme's surprise list (this would be duplicated across each `.rs` file).

**Comment by CaspianA1** [comment]
To implement this, surprises will need to depend on not just an image path, but a full `TextureCreationInfo` struct as their image data source (a DJ tip would then be a text texture).

For various technical reasons, this will mean that `TextureCreationInfo` will need to be hashable (which I actually just implemented [here](https://github.com/WBOR-91-1-FM/wbor-studio-dashboard/commit/e4b9c50fbda6ccec87493361855917ed297360c3))!

So a major roadblock for implementing this is now completed :)

**Comment by mdrxy** [comment]
More thoughts on this: we could use a local SQLite database to track whether certain people (via their Spinitron persona IDs) have seen a given tip, as to not show duplicate ones after `x` number of appearances.

**Comment by CaspianA1** [comment]
Good idea - might be easier to keep this in-memory though (assuming no restarts or crashes).

Or if we make the number of tips quite high, and also somewhat infrequent, the chance of repetition would decrease even more.
